TESTS/mbedmicro-rtos-mbed/timer/main.cpp

```diff
@@ -0,0 +1,872 @@
+    /* Start the timer. */
+    p_timer->start();
+
+    /* -- Simulate that current time is equal to 1 ms -- */
+    curr_ticker_us_val = 1000;
+
+    /* Stop the timer. */
+    p_timer->stop();
+
+    /* Read timer value using each read function. */
+    read_val = p_timer->read();
+    read_ms_val = p_timer->read_ms();
+    read_us_val = p_timer->read_us();
+    read_hr_us_val = p_timer->read_high_resolution_us();
+
+    /* Check results - 1 ms has elapsed. */
+    TEST_ASSERT_EQUAL_FLOAT(0.001000f, read_val);
+    TEST_ASSERT_EQUAL(1, read_ms_val);
+    TEST_ASSERT_EQUAL(1000, read_us_val);
+    TEST_ASSERT_EQUAL(1000, read_hr_us_val);
+
+    /* Start the timer. */
+    p_timer->start();
+
+    /* -- Simulate that current time is equal to 125 ms -- */
+    curr_ticker_us_val = 125000;
+
+    /* Stop the timer. */
+    p_timer->stop();
+
+    /* Read timer value using each read function. */
+    read_val = p_timer->read();
+    read_ms_val = p_timer->read_ms();
+    read_us_val = p_timer->read_us();
+    read_hr_us_val = p_timer->read_high_resolution_us();
+
+    /* Check results - 125 ms have elapsed. */
+    TEST_ASSERT_EQUAL_FLOAT(0.125000f, read_val);
+    TEST_ASSERT_EQUAL(125, read_ms_val);
+    TEST_ASSERT_EQUAL(125000, read_us_val);
+    TEST_ASSERT_EQUAL(125000, read_hr_us_val);
+
+    /* Start the timer. */
+    p_timer->start();
+
+    /* -- Simulate that current time is equal to 1 s -- */
+    curr_ticker_us_val = 1000000;
+
+    /* Stop the timer. */
+    p_timer->stop();
+
+    /* Read timer value using each read function. */
+    read_val = p_timer->read();
+    read_ms_val = p_timer->read_ms();
+    read_us_val = p_timer->read_us();
+    read_hr_us_val = p_timer->read_high_resolution_us();
+
+    /* Check results - 1 s has elapsed. */
+    TEST_ASSERT_EQUAL_FLOAT(1.000000f, read_val);
+    TEST_ASSERT_EQUAL(1000, read_ms_val);
+    TEST_ASSERT_EQUAL(1000000, read_us_val);
+    TEST_ASSERT_EQUAL(1000000, read_hr_us_val);
+
+    /* Start the timer. */
+    p_timer->start();
+
+    /* -- Simulate that current time is equal to 125 s -- */
+    curr_ticker_us_val = 125000000;
+
+    /* Stop the timer. */
+    p_timer->stop();
+
+    /* Read timer value using each read function. */
+    read_val = p_timer->read();
+    read_ms_val = p_timer->read_ms();
+    read_us_val = p_timer->read_us();
+    read_hr_us_val = p_timer->read_high_resolution_us();
+
+    /* Check results - 125 s have elapsed. */
+    TEST_ASSERT_EQUAL_FLOAT(125.000000f, read_val);
+    TEST_ASSERT_EQUAL(125000, read_ms_val);
+    TEST_ASSERT_EQUAL(125000000, read_us_val);
+    TEST_ASSERT_EQUAL(125000000, read_hr_us_val);
+
+    /* Start the timer. */
+    p_timer->start();
+
+    /* -- Simulate that current time is equal to MAX_INT_32 us -- */
+
+    /* Note that timers are based on 32-bit int microsecond counters,
+     * so can only time up to a maximum of 2^31-1 microseconds i.e.
+     * 2147483647 us (about 35 minutes). */
+    curr_ticker_us_val = 2147483647;
```

Comment:
Serious question, what is supposed to happens when curr_ticker_us_val.read returns a value outside the [0 : MAX_INT] range ?

Reply:
@bulislaw Could you clarify the behavior of the Timer class when the time measured exceeds the value returned ? 

There is a warning in the handbook but nothing in the class declaration. The range returned by `read`, `read_ms`, `read_us` and `read_high_resolution_us` is also unspecified. According to the warning in the handbook, the maximum value returned by `read_ms` will be `2147483` which is not very intuitive.

Reply:
I'm not sure, best would be to check how it actually works. Confirm it makes sense and update the docs/handbook.

Reply:
@bulislaw @pan- I created test based on documentation provided in handbook. So the boundary case for max measured time is MAX_INT. I agree that values returned by the read* functions are not very intuitive. Currently when measured time is equal to MAX_INT (2147483647) read* functions return the following values:
read(): 2147.483643 - ok
read_ms(): 2147483 - ok
read_us(): 2147483647 - ok
read_high_resolution_us(): 2147483647 - ok
When measured time is equal to MAX_INT+1 (2147483648), then we have the following results:
read(): -2147.483643 - wrong
read_ms(): 2147483 - ok
read_us(): -2147483648 - wrong
read_high_resolution_us(): 2147483648 - ok.

It is very strange that returned time can be negative, but warning in handbook informs about Timer time measurement limitations. On the other hand maybe we could modify Timer::read_us()
and Timer::read_ms() function declarations and change returned type to unsigned int. This way measured time limit would be increased and we would also rid of negative results. 
In such case handbook and documents need to be also updated.